### PR TITLE
SCP-2293: use PrimArray instead of STRef

### PR DESCRIPTION
**Pull request by michaelpj**
Surprisingly effective!
```
crowdfunding/1            1.467 ms       1.347 ms       -8.2%
crowdfunding/2            1.462 ms       1.344 ms       -8.1%
crowdfunding/3            1.457 ms       1.340 ms       -8.0%
crowdfunding/4            706.4 μs       655.3 μs       -7.2%
crowdfunding/5            709.5 μs       654.0 μs       -7.8%
future/1                  833.7 μs       762.6 μs       -8.5%
future/2                  1.729 ms       1.564 ms       -9.5%
future/3                  1.727 ms       1.567 ms       -9.3%
future/4                  2.158 ms       1.960 ms       -9.2%
future/5                  2.642 ms       2.395 ms       -9.3%
future/6                  2.201 ms       2.017 ms       -8.4%
future/7                  2.638 ms       2.392 ms       -9.3%
multisigSM/1              1.540 ms       1.425 ms       -7.5%
multisigSM/2              1.561 ms       1.448 ms       -7.2%
multisigSM/3              1.583 ms       1.474 ms       -6.9%
multisigSM/4              1.601 ms       1.492 ms       -6.8%
multisigSM/5              1.903 ms       1.761 ms       -7.5%
multisigSM/6              1.533 ms       1.431 ms       -6.7%
multisigSM/7              1.558 ms       1.455 ms       -6.6%
multisigSM/8              1.591 ms       1.474 ms       -7.4%
multisigSM/9              1.613 ms       1.491 ms       -7.6%
multisigSM/10             1.916 ms       1.756 ms       -8.4%
vesting/1                 1.490 ms       1.364 ms       -8.5%
vesting/2                 1.298 ms       1.189 ms       -8.4%
vesting/3                 1.489 ms       1.362 ms       -8.5%
marlowe/trustfund/1       4.134 ms       3.748 ms       -9.3%
marlowe/trustfund/2       3.031 ms       2.781 ms       -8.2%
marlowe/zerocoupon/1      4.224 ms       3.850 ms       -8.9%
marlowe/zerocoupon/2      2.673 ms       2.472 ms       -7.5%
```

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/ExBudgetMode.hs [hunk 1]
```diff
@@ -106,25 +110,38 @@ instance Pretty RestrictingSt where
...
+    -- annoying to specify the monad, since it refers to the 's' which is not in scope.
+    -- So we use a partial type signature.
+    (ref :: MutablePrimArray _ SatInt) <- newPrimArray 2
+    let
+        cpuIx = 0
+        memIx = 1
+        readCpu = coerce <$> readPrimArray ref cpuIx
+        writeCpu cpu = writePrimArray ref cpuIx $! coerce cpu
+        readMem = coerce <$> readPrimArray ref memIx
+        writeMem mem = writePrimArray ref memIx $! coerce mem
```
Comment (effectfully) [inline plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/ExBudgetMode.hs:125]: We don't need the bangs anymore given that you can't put a thunk into some unboxed place, right?

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/ExBudgetMode.hs [hunk 2]
```diff
@@ -106,25 +110,38 @@ instance Pretty RestrictingSt where
...
-    memRef <- newSTRef memInit
-    let spend _ (ExBudget cpuToSpend memToSpend) = do
-            cpuLeft <- readSTRef cpuRef
-            memLeft <- readSTRef memRef
+    -- We keep the counters in a PrimArray. This is better than an STRef since it stores its contents unboxed.
+    --
+    -- If we don't specify the element type then GHC has difficulty inferring it, but it's
+    -- annoying to specify the monad, since it refers to the 's' which is not in scope.
+    -- So we use a partial type signature.
+    (ref :: MutablePrimArray _ SatInt) <- newPrimArray 2
```
Comment (effectfully) [inline plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/ExBudgetMode.hs:118]: I'd write that as `ref <- newPrimArray @_ @SatInt` but whatever.

**Review by effectfully — APPROVED** [review]
Wow!

**Comment by kwxm** [comment]
Ah! We've broken the 4ms barrier.  Next stop 3ms.

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/ExBudgetMode.hs (hunk 1 above)
Comment (michaelpj) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/ExBudgetMode.hs:125]: Good point.

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/ExBudgetMode.hs (hunk 2 above)
Comment (michaelpj) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/ExBudgetMode.hs:118]: Ah yeah, I always forget about `@_`

**Review by kwxm — COMMENTED** [review]
Looks good.  If I wanted to experiment with removing memory costs, what would I do?  I could use an array of length one, but is there something simpler?

**Review by kwxm — APPROVED** [review]
Oops, I meant to approve that.

**Comment by michaelpj** [comment]
> Looks good. If I wanted to experiment with removing memory costs, what would I do? I could use an array of length one, but is there something simpler?

I think just an array of length 1 would be the thing. As far as I know, there is no unboxed `STRef`, so we'd have to fake it with a small `PrimArray` regardless.

**Comment by michaelpj** [comment]
Hm, we could use https://hackage.haskell.org/package/unboxed-ref-0.4.0.0/docs/Data-STRef-Unboxed.html, I guess?

**Comment by michaelpj** [comment]
(which is the same thing, just packaged up)

**Pull request merged**